File: server/internal/models/trip.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type TripStatus string

const (
	TripStatusActive    TripStatus = "active"
	TripStatusNotified  TripStatus = "notified"
	TripStatusCancelled TripStatus = "cancelled"
	TripStatusExpired   TripStatus = "expired"
)

type Trip struct {
	ID               uuid.UUID  `json:"id"`
	UserID           uuid.UUID  `json:"user_id"`
	OriginLat        float64    `json:"origin_lat"`
	OriginLng        float64    `json:"origin_lng"`
	OriginName       string     `json:"origin_name"`
	DestinationLat   float64    `json:"destination_lat"`
	DestinationLng   float64    `json:"destination_lng"`
	DestinationName  string     `json:"destination_name"`
	DesiredArrivalAt time.Time  `json:"desired_arrival_at"`
	WarningMinutes   int        `json:"warning_minutes"`
	Status           TripStatus `json:"status"`
	// Latest ETA from Google Maps in seconds.
	LatestETASeconds int `json:"latest_eta_seconds"`
	// Previous ETA used for hysteresis check.
	PrevETASeconds int `json:"-"`
	// Consecutive polls where ETA was stable (within ±3 min of prev).
	StableETACount     int        `json:"-"`
	NextPollAt         time.Time  `json:"next_poll_at"`
	NotificationSentAt *time.Time `json:"notification_sent_at,omitempty"`
	CreatedAt          time.Time  `json:"created_at"`
	UpdatedAt          time.Time  `json:"updated_at"`
}

// ShouldLeaveAt computes the time the user should depart.
func (t *Trip) ShouldLeaveAt() time.Time {
	eta := time.Duration(t.LatestETASeconds) * time.Second
	warning := time.Duration(t.WarningMinutes) * time.Minute
	return t.DesiredArrivalAt.Add(-eta - warning)
}

File: onTime/onTime/Core/Models/Trip.swift
import Foundation
import CoreLocation

struct Trip: Codable, Identifiable, Hashable {
    let id: UUID
    let userId: UUID
    let originLat: Double
    let originLng: Double
    let originName: String
    let destinationLat: Double
    let destinationLng: Double
    let destinationName: String
    let desiredArrivalAt: Date
    let warningMinutes: Int
    let status: Status
    let latestEtaSeconds: Int
    let nextPollAt: Date
    let notificationSentAt: Date?
    let createdAt: Date
    let updatedAt: Date

    enum Status: String, Codable {
        case active, notified, cancelled, expired
    }

    var isActive: Bool { status == .active }

    var shouldLeaveAt: Date {
        desiredArrivalAt.addingTimeInterval(-Double(latestEtaSeconds + warningMinutes * 60))
    }

    var origin: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: originLat, longitude: originLng)
    }

    var destination: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: destinationLat, longitude: destinationLng)
    }

    var hasOrigin: Bool { originLat != 0 || originLng != 0 }

    enum CodingKeys: String, CodingKey {
        case id, status
        case userId = "user_id"
        case originLat = "origin_lat"
        case originLng = "origin_lng"
        case originName = "origin_name"
        case destinationLat = "destination_lat"
        case destinationLng = "destination_lng"
        case destinationName = "destination_name"
        case desiredArrivalAt = "desired_arrival_at"
        case warningMinutes = "warning_minutes"
        case latestEtaSeconds = "latest_eta_seconds"
        case nextPollAt = "next_poll_at"
        case notificationSentAt = "notification_sent_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}
